crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,801 @@
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_canonicalize(
+    obs: *const SparseObservable,
+    tol: f64,
+) -> *mut SparseObservable {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let result = obs.canonicalize(tol);
+    Box::into_raw(Box::new(result))
+}
+
+/// @ingroup QkSparseObservable
+/// Copy the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to a copy of the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *original = qk_obs_identity(100);
+///     QkSparseObservable *copied = qk_obs_copy(original);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_copy(obs: *const SparseObservable) -> *mut SparseObservable {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let copied = obs.clone();
+    Box::into_raw(Box::new(copied))
+}
+
+/// @ingroup QkSparseObservable
+/// Compare two observables for equality.
+///
+/// Note that this does not compare mathematical equality, but data equality. This means
+/// that two observables might represent the same observable but not compare as equal.
+///
+/// @param obs A pointer to one observable.
+/// @param other A pointer to another observable.
+///
+/// @return ``true`` if the observables are equal, ``false`` otherwise.
+///
+/// # Example
+///
+///     QkSparseObservable *observable = qk_obs_identity(100);
+///     QkSparseObservable *other = qk_obs_identity(100);
+///     bool are_equal = qk_obs_equal(observable, other);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` or ``other`` are not valid, non-null pointers to
+/// ``QkSparseObservable``\ s.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_equal(
+    obs: *const SparseObservable,
+    other: *const SparseObservable,
+) -> bool {
+    // SAFETY: Per documentation, the pointers are non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let other = unsafe { const_ptr_as_ref(other) };
+
+    obs.eq(other)
+}
+
+/// @ingroup QkSparseObservable
+/// Print the observable.
+///
+/// @param obs A pointer to the ``SparseObservable`` to print.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     qk_obs_print(obs);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_print(obs: *const SparseObservable) {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    println!("{:?}", obs);
+}
+
+/// @ingroup QkSparseTerm
+/// Print the sparse term.
```

Comment:
I would probably add:

```suggestion
/// Print the sparse term to stdout.
```

To make it clear that it's printing to stdout.

Reply:
We should most likely remove this function and only have it return a string - having multiple places access `stdout` through different runtimes is going to be a nightmare of locking and buffering, I'd think.  Unless you immediately know best practices here, might be easiest to just return allocated strings, and add to the API later with some stream printer, if we really need it.

Reply:
Yeah this bit maybe needs some updating... we currently (.. or ever?) cannot coherently print from Rust and C: interleaving print statements like
```
printf("observable a:\n");
qk_obs_print(a);
printf("observable b:\n");
qk_obs_print(b);
```
prints first all Rust messages and then C (or possibly the other way around I can't quite remember right now). So with @jakelishman's comments from above it might be the better solution to just have a function to get the C-string representation of the observable, which can be printed with `printf` from C

Reply:
I pushed up a commit modifying this methods to return strings: https://github.com/Qiskit/qiskit/pull/13694/commits/603ae56df0038d97b371ea720eed75226c48c9b4 someone else should validate they look ok though.

Reply:
This documentation needs changing now because the function doesn't print any more haha

Reply:
Updated in 624c3dd